---
title: "sbl: an R package implementing sparse Bayesian learning method for QTL mapping and GWAS"
author: |-
  Meiyue Wang and Shizhong Xu
 
  Department of Botany and Plant Sciences, University of California, Riverside
date: "October 6, 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

Single marker model that detecting one locus at a time encounter many problems in 
genome-wide association study (GWAS) of quantitative traits, such as involving large 
matrix calculation and inflation of significant markers caused by linkage disequilibrium 
 (LD). Such problems can be solved by multiple locus model which includes all markers 
in the same model with effects being estimated simultaneously. The Sparse Bayesian 
learning method (SBL), implemented in `sbl` package, is a multiple locus model that can 
handle extremely large sample size (>100,000) and computationally efficient. 


# `sbl` package installation

`sbl` can be downloaded and installed locally. The download link is
[here](https://github.com/mwang024/sbl_0.1.0.tar.gz).

```{r installation, eval=FALSE, message=FALSE, warning=FALSE}
install.packages('sbl_0.1.0.tar.gz', repos=NULL, type='source')
```


# SBL for QTL mapping and GWAS

The usage of `sbl` to perform QTL mapping and GWAS is very simple:


1. Load input data
    + A vector of response variables (observations of the trait).
    + A design matrix for fixed effects, which represents non-genetic factors that 
contribute to the phenotypic variance. This can be intercept only.
    + A matrix storing genotype indicators. The original three genotypes 
  ($a_1a_1$ $a_1a_2$ $a_2a_2$) should be coded to numeric indicator as (-1 0 1) 
    or (0 1 2).

```{r load data, message=FALSE, warning=FALSE}
library('sbl')

# load example data
data(phe)
data(intercept)
data(gen)
```


2. Call `sblas()` function to perform regression and detect significant markers.

```{r minimal invocation, message=FALSE, warning=FALSE,}
# A minimal invocation of 'sblas()' function looks like: 
fit1<-sblas(x = intercept, y = phe, z = gen)

# Restuls of markers surrounding the first simulated QTL in the example data
fit1$blup[c(19:23),]
```


Users can arbitrarily set the value of `t` between [0, 2] to control the sparseness of model, 
default is -1. 

```{r hyper parameter, message=FALSE, warning=FALSE}
# Set t = 0 to make the model most sparse 
fit2<-sblas(x = intercept, y = phe, z = gen, t = 0)
fit2$parm

# Set t = -2 to make the model least sparse 
fit3<-sblas(x = intercept, y = phe, z = gen, t = -2)
fit3$parm
```


`max.iter` and `min.err` are two threshold to stop the program when either of them
is met. 
`max.iter` defines the maximum number of iterations that the program is 
allowed to run, default is 200; 
`min.err` defines the minimum threshold of mean squared error of random effects 
estimated from the current and the previous iteration, default is 1e-6.

```{r iteration, message=FALSE, warning=FALSE}
# Set max.iter and min.err to control the convergence of the model 
fit4<-sblas(x = intercept, y = phe, z = gen, t = -1, max.iter = 300, min.err = 1e-8)
fit4$parm
```

